### DISCUSSION TITLE
Create a CommandOption that is required

### DISCUSSION THREAD

**Discussion by MrHinsh** [spectreconsole/spectre.console#1568] [Q&A] [answered]
`CommandOption` has a field that is `ValueIsOptional` which seams to be set out of the string template that is passed in.

![image](https://github.com/spectreconsole/spectre.console/assets/5205575/550fe332-0b36-48b5-bd1a-d96071e8c37a)

How do I set `ValueIsOptional` to `false`?

**Comment by patriksvensson** [comment]
~~The parameter needs to be of the type `FlagValue<T>`: https://spectreconsole.net/api/spectre.console.cli/flagvalue_1/#typeparam-T~~

**Reply by MrHinsh** [reply]
I have no idea what that means @patriksvensson. And I don't see any help on that link as to what the `template` test would look like.

What would it look like on:
`[CommandOption("-p|--projectId")]`

**Reply by patriksvensson** [reply, marked as answer]
Sorry, I misunderstood. All options are optional by default. If you want to enforce an option and make the option required, the best way is to override the `Validate(CommandContext context, TSettings settings)` on either the setting itself or the command.

```csharp
public class MyCommand : Command<MyCommand.Settings>
{
    public sealed class Settings : CommandSettings
    {
        [CommandOption("-p|--projectId")]
        public string? ProjectId { get; set; }

        // You can validate the settings directly
        public override Validate(CommandContext context) 
        {
            if (ProjectID == null) {
                return ValidationResult.Error("Project ID id required");
            }

            return ValidationResult.Success();
        }
    }

    public override int Execute(CommandContext context, Settings settings)
    {
        return 0;
    }

    // OR, you can perform validation in the command
    // might be more usable if you require external services etc
    public override Validate(CommandContext context, Settings settings) 
    {
        if (settings.ProjectID == null) {
            return ValidationResult.Error("Project ID id required");
        }

        return ValidationResult.Success();
    }
}
```

Not sure if the code above compiles, but that should solve your problem.
